fuzzers/035-iob-ilogic/top.py

```diff
@@ -0,0 +1,442 @@
+    iobdelay = random.choice((
+        'NONE',
+        'BOTH',
+        'IBUF',
+        'IFD',
+    ))
+
+    p['IOBDELAY'] = verilog.quote(iobdelay)
+    p['INIT_Q1'] = random.randint(0, 1)
+    p['INIT_Q2'] = random.randint(0, 1)
+    p['INIT_Q3'] = random.randint(0, 1)
+    p['INIT_Q4'] = random.randint(0, 1)
+
+    p['SRVAL_Q1'] = random.randint(0, 1)
+    p['SRVAL_Q2'] = random.randint(0, 1)
+    p['SRVAL_Q3'] = random.randint(0, 1)
+    p['SRVAL_Q4'] = random.randint(0, 1)
+    p['NUM_CE'] = random.randint(1, 2)
+
+    p['IS_CLK_INVERTED'] = random.randint(0, 1)
+    p['IS_CLKB_INVERTED'] = random.randint(0, 1)
+    p['IS_OCLK_INVERTED'] = random.randint(0, 1)
+    p['IS_D_INVERTED'] = random.randint(0, 1)
+    p['INTERFACE_TYPE'] = verilog.quote(
+        random.choice(
+            (
+                'MEMORY',
+                'MEMORY_DDR3',
+                'MEMORY_QDR',
+                'NETWORKING',
+                'OVERSAMPLE',
+            )))
+    p['DATA_RATE'] = verilog.quote(random.choice((
+        'SDR',
+        'DDR',
+    )))
+    if verilog.unquote(p['DATA_RATE']) == 'SDR':
+        data_widths = [2, 3, 4, 5, 6, 7, 8]
+    else:
+        data_widths = [4, 6, 8]
+
+    p['DATA_WIDTH'] = random.choice(data_widths)
+    p['SERDES_MODE'] = verilog.quote(random.choice(('MASTER', 'SLAVE')))
+
+    use_delay = iobdelay != 'NONE'
+
+    if iobdelay == 'NONE':
+        p['mux_config'] = 'direct'
+        p['iddr_mux_config'] = 'direct'
+    elif iobdelay == 'BOTH':
+        p['mux_config'] = 'idelay'
+        p['iddr_mux_config'] = 'idelay'
+    elif iobdelay == 'IBUF':
+        p['mux_config'] = 'idelay'
+        p['iddr_mux_config'] = 'direct'
+    elif iobdelay == 'IFD':
+        p['mux_config'] = 'direct'
+        p['iddr_mux_config'] = 'idelay'
+
+    p['OFB_USED'] = verilog.quote(random.choice(('TRUE', 'FALSE')))
+    p['DYN_CLKDIV_INV_EN'] = verilog.quote(random.choice(('TRUE', 'FALSE')))
+    p['DYN_CLK_INV_EN'] = verilog.quote(random.choice(('TRUE', 'FALSE')))
+
+    if use_delay:
+        print(
+            """
+    wire idelay_{site};
+
+    (* KEEP, DONT_TOUCH, LOC = "{idelay_loc}" *)
+    IDELAYE2 #(
+    ) idelay_site_{site} (
+        .IDATAIN({iwire}),
+        .DATAOUT(idelay_{site})
+        );""".format(**p),
+            file=connects)
+
+        p['ddly_connection'] = '.DDLY(idelay_{site}),'.format(**p)
+    else:
+        p['ddly_connection'] = ''
+
+    if verilog.unquote(p['OFB_USED']) == 'TRUE':
+        p['ODATA_RATE'] = verilog.quote(random.choice((
+            'SDR',
+            'DDR',
+        )))
+        if verilog.unquote(p['ODATA_RATE']) == 'SDR':
+            data_widths = [2, 3, 4, 5, 6, 7, 8]
+        else:
+            data_widths = [4, 6, 8]
+
+        p['ODATA_WIDTH'] = random.choice(data_widths)
+        p['OSERDES_MODE'] = verilog.quote(random.choice(('MASTER', 'SLAVE')))
+
+        if p['ODATA_WIDTH'] > 4 or verilog.unquote(p['ODATA_RATE']) == 'SDR':
```

Comment:
this seems to be invalid. According to ug471 [(v1.10) May 8, 2018] `TRISTATE_WIDTH = 4` only when RATE is DDR and ODATA_WIDTH = 4. In this case we'll set TRISTATE_WIDTH = 4 for all ODATA_WITH's > 4 (but not 4).

Reply:
Agreed, fixed.